Superconductivity critical temperature prediction 

Import all necessary libraries for model building

In [2]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.preprocessing import LabelEncoder,OneHotEncoder,MinMaxScaler
#from sklearn.scale import MinMaxScaler


Read the data set 

In [3]:
data = pd.read_csv("D:\\sai data science\\data science practise\\Data sets\\superconduct\\train.csv")

In [4]:
data.head(3)

number_of_elements  mean_atomic_mass  wtd_mean_atomic_mass  \
0                   4         88.944468             57.862692   
1                   5         92.729214             58.518416   
2                   4         88.944468             57.885242   

   gmean_atomic_mass  wtd_gmean_atomic_mass  entropy_atomic_mass  \
0          66.361592              36.116612             1.181795   
1          73.132787              36.396602             1.449309   
2          66.361592              36.122509             1.181795   

   wtd_entropy_atomic_mass  range_atomic_mass  wtd_range_atomic_mass  \
0                 1.062396          122.90607              31.794921   
1                 1.057755          122.90607              36.161939   
2                 0.975980          122.90607              35.741099   

   std_atomic_mass  ...  wtd_mean_Valence  gmean_Valence  wtd_gmean_Valence  \
0        51.968828  ...          2.257143       2.213364           2.219783   
1        47.094633  ...          2.257143       1.888175           2.210679   
2        51.968828  ...          2.271429       2.213364           2.232679   

   entropy_Valence  wtd_entropy_Valence  range_Valence  wtd_range_Valence  \
0         1.368922             1.066221              1           1.085714   
1         1.557113             1.047221              2           1.128571   
2         1.368922             1.029175              1           1.114286   

   std_Valence  wtd_std_Valence  critical_temp  
0     0.433013         0.437059           29.0  
1     0.632456         0.468606           26.0  
2     0.433013         0.444697           19.0  

[3 rows x 82 columns]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21263 entries, 0 to 21262
Data columns (total 82 columns):
number_of_elements                 21263 non-null int64
mean_atomic_mass                   21263 non-null float64
wtd_mean_atomic_mass               21263 non-null float64
gmean_atomic_mass                  21263 non-null float64
wtd_gmean_atomic_mass              21263 non-null float64
entropy_atomic_mass                21263 non-null float64
wtd_entropy_atomic_mass            21263 non-null float64
range_atomic_mass                  21263 non-null float64
wtd_range_atomic_mass              21263 non-null float64
std_atomic_mass                    21263 non-null float64
wtd_std_atomic_mass                21263 non-null float64
mean_fie                           21263 non-null float64
wtd_mean_fie                       21263 non-null float64
gmean_fie                          21263 non-null float64
wtd_gmean_fie                      21263 non-null float64
entropy_fie            

One thing is clear that there is no null values in the dataset as all columns are having all 21263 notnull values

In [6]:
y = data['critical_temp']
data.drop(['critical_temp'],axis = 1,inplace=True)

In [7]:
data.columns

Index(['number_of_elements', 'mean_atomic_mass', 'wtd_mean_atomic_mass',
       'gmean_atomic_mass', 'wtd_gmean_atomic_mass', 'entropy_atomic_mass',
       'wtd_entropy_atomic_mass', 'range_atomic_mass', 'wtd_range_atomic_mass',
       'std_atomic_mass', 'wtd_std_atomic_mass', 'mean_fie', 'wtd_mean_fie',
       'gmean_fie', 'wtd_gmean_fie', 'entropy_fie', 'wtd_entropy_fie',
       'range_fie', 'wtd_range_fie', 'std_fie', 'wtd_std_fie',
       'mean_atomic_radius', 'wtd_mean_atomic_radius', 'gmean_atomic_radius',
       'wtd_gmean_atomic_radius', 'entropy_atomic_radius',
       'wtd_entropy_atomic_radius', 'range_atomic_radius',
       'wtd_range_atomic_radius', 'std_atomic_radius', 'wtd_std_atomic_radius',
       'mean_Density', 'wtd_mean_Density', 'gmean_Density',
       'wtd_gmean_Density', 'entropy_Density', 'wtd_entropy_Density',
       'range_Density', 'wtd_range_Density', 'std_Density', 'wtd_std_Density',
       'mean_ElectronAffinity', 'wtd_mean_ElectronAffinity',
       'gmean_

In [15]:
data.corr()

number_of_elements  mean_atomic_mass  \
number_of_elements                         1.000000         -0.141923   
mean_atomic_mass                          -0.141923          1.000000   
wtd_mean_atomic_mass                      -0.353064          0.815977   
gmean_atomic_mass                         -0.292969          0.940298   
wtd_gmean_atomic_mass                     -0.454525          0.745841   
entropy_atomic_mass                        0.939304         -0.104000   
wtd_entropy_atomic_mass                    0.881845         -0.097609   
range_atomic_mass                          0.682777          0.125659   
wtd_range_atomic_mass                     -0.320293          0.446225   
std_atomic_mass                            0.513998          0.196460   
wtd_std_atomic_mass                        0.546391          0.130675   
mean_fie                                   0.167451         -0.285782   
wtd_mean_fie                               0.484445         -0.222097   
gmean_fie                                  0.024229         -0.240565   
wtd_gmean_fie                              0.424152         -0.219381   
entropy_fie                                0.973195         -0.166935   
wtd_entropy_fie                            0.719209         -0.163565   
range_fie                                  0.781227         -0.255628   
wtd_range_fie                              0.329624         -0.080545   
std_fie                                    0.674005         -0.276561   
wtd_std_fie                                0.717831         -0.222812   
mean_atomic_radius                        -0.001389          0.497664   
wtd_mean_atomic_radius                    -0.422144          0.376760   
gmean_atomic_radius                       -0.240444          0.561061   
wtd_gmean_atomic_radius                   -0.518256          0.359894   
entropy_atomic_radius                      0.972245         -0.140034   
wtd_entropy_atomic_radius                  0.904121         -0.147604   
range_atomic_radius                        0.768060         -0.270695   
wtd_range_atomic_radius                   -0.371350          0.141100   
std_atomic_radius                          0.624810         -0.326403   
...                                             ...               ...   
mean_FusionHeat                           -0.437624         -0.137669   
wtd_mean_FusionHeat                       -0.449272         -0.135429   
gmean_FusionHeat                          -0.514252          0.014818   
wtd_gmean_FusionHeat                      -0.519109         -0.043003   
entropy_FusionHeat                         0.900759         -0.008499   
wtd_entropy_FusionHeat                     0.860479         -0.028541   
range_FusionHeat                           0.005734         -0.347582   
wtd_range_FusionHeat                      -0.371788         -0.167528   
std_FusionHeat                            -0.113361         -0.337969   
wtd_std_FusionHeat                        -0.074796         -0.335778   
mean_ThermalConductivity                   0.227656         -0.158266   
wtd_mean_ThermalConductivity               0.206069         -0.065989   
gmean_ThermalConductivity                 -0.485324          0.006004   
wtd_gmean_ThermalConductivity             -0.469206          0.056394   
entropy_ThermalConductivity                0.501871         -0.100077   
wtd_entropy_ThermalConductivity            0.207065         -0.098221   
range_ThermalConductivity                  0.696060         -0.114538   
wtd_range_ThermalConductivity              0.316772         -0.027790   
std_ThermalConductivity                    0.602018         -0.110658   
wtd_std_ThermalConductivity                0.665580         -0.110856   
mean_Valence                              -0.609412          0.374099   
wtd_mean_Valence                          -0.648551          0.304683   
gmean_Valence                             -0.618512          0.392153   
wtd_gmean_Valence 

Just directly built the model and will compare the results of preprocessed data and plain data 

Divide the data set into train and test 

In [9]:
x_train,x_test,y_train,y_test = train_test_split(data,y,test_size = 0.2)
x_train.columns
y_train


1206      77.000
3706      58.000
13164      0.500
20954     20.300
11946      2.680
11371      4.000
11235      1.950
18812      1.730
10451      6.200
8106      34.000
13505      7.500
12673      8.600
17999      5.000
10132      2.620
1618      62.000
20793      3.500
5156      18.000
11584      9.600
18626     31.200
20186      5.000
11086      6.590
5690      76.000
13554     13.800
18569      6.010
16681      4.200
4842      38.000
2636      22.000
17653    103.500
11955      2.240
964        2.900
          ...   
5968     121.000
19275     18.600
5727      42.000
11845     17.000
17779     94.000
15857      8.320
10157      9.400
2858      87.000
7749      42.000
11093      1.800
18833     14.900
5917      91.000
293       81.000
18131     17.000
17451     17.300
17739     90.520
4473      30.700
2455      40.000
2359      85.000
7479      80.000
6323       2.650
3247       0.001
8493      13.800
2360      80.000
1511      66.000
4330      82.000
14919      0.840
1956      82.5

# 1.Trying with RandomForestRegressor

In [10]:
from sklearn.ensemble import RandomForestRegressor
from math import sqrt
rfr = RandomForestRegressor()
rfrmodel1 = rfr.fit(x_train,y_train)
rfrmodel1
#make predictions by this model
y_pred = rfrmodel1.predict(x_test)
#checking the model performance by MAE,RMSE,MSE
print("MAE : ", metrics.mean_absolute_error(y_test,y_pred))
print("RMSE : ", sqrt(metrics.mean_squared_error(y_test,y_pred)))
print("MSE : ", metrics.mean_squared_error(y_test,y_pred))

D:\sai data science\Installations\Anaconda\Anaconda new installation\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


MAE :  5.392033083915986
RMSE :  9.78241905840192
MSE :  95.69572263418512


# 2.Linear regression model 

In [11]:
from sklearn.linear_model import LinearRegression
lr= LinearRegression()
lrmodel1 = lr.fit(x_train,y_train)
#making predictions by lrmodel1
y_predlr1 = lrmodel1.predict(x_test) 
#checking the model performance by MAE,RMSE,MSE
print("MAE : ", metrics.mean_absolute_error(y_test,y_predlr1))
print("RMSE : ", sqrt(metrics.mean_squared_error(y_test,y_predlr1)))
print("MSE : ", metrics.mean_squared_error(y_test,y_predlr1))

MAE :  13.402247851823326
RMSE :  17.7130286150829
MSE :  313.75138271874573


preprocessing the data with minmaxscaler,get dummies for categorical variables and we will see if there is any improvement in the model performance 

In [12]:
# as we know from data.info()  that there are three categorical variables,we need to convert them into binary variables
noe = pd.get_dummies(data['number_of_elements'])
rar = pd.get_dummies(data['range_atomic_radius'])
rv = pd.get_dummies(data['range_Valence'])
data2 = data
# wealready created dummy variables for those three variables,so we can drop the original columns from the dataframe
data2.drop(['number_of_elements','range_atomic_radius','range_Valence'],axis =1,inplace= True)
#data2 = pd.DataFrame(data2)

In [14]:
#add those three dummies variables to our data set
data3 = pd.concat([data2,noe,rv,rar],axis=1)
type(data2)
type(noe)
type(rar)
type(rv)

pandas.core.frame.DataFrame

In [16]:
total_rows = len(data3.axes[0])
total_cols = len(data3.axes[1])
total_rows
total_cols

289

In [17]:
#appluy minmax scaling 
cols = data3.columns
mxm = MinMaxScaler()
mxm_data = pd.DataFrame(mxm.fit_transform(data3),columns = cols)


In [18]:
#make predictions using new data which is preprocessed and will compare the results
x_tr,x_te,y_tr,y_te = train_test_split(mxm_data,y,test_size = 0.2)

# Random forest regressor model 2

In [19]:
rfr2 = RandomForestRegressor()
rfrmodel2 = rfr2.fit(x_tr,y_tr)
y_pred2 = rfrmodel2.predict(x_te)
#checking the model performance by MAE,RMSE,MSE
print("MAE : ", metrics.mean_absolute_error(y_te,y_pred2))
print("RMSE : ", sqrt(metrics.mean_squared_error(y_te,y_pred2)))
print("MSE : ", metrics.mean_squared_error(y_te,y_pred2))

D:\sai data science\Installations\Anaconda\Anaconda new installation\lib\site-packages\sklearn\ensemble\forest.py:245: FutureWarning: The default value of n_estimators will change from 10 in version 0.20 to 100 in 0.22.
  "10 in version 0.20 to 100 in 0.22.", FutureWarning)


MAE :  5.268089227612972
RMSE :  9.417841975251303
MSE :  88.69574747080537


In [20]:
# we saw that there is little bit improvement in the model performance , now try to tune the hyper parameters of random forest regressor 

# Linear regression model2

In [21]:
lr2 = LinearRegression()
lrmodel2 = lr2.fit(x_tr,y_tr)
y_predlr2 = lrmodel2.predict(x_te)
#checking the model performance by MAE,RMSE,MSE
print("MAE : ", metrics.mean_absolute_error(y_te,y_predlr2))
print("RMSE : ", sqrt(metrics.mean_squared_error(y_te,y_predlr2)))
print("MSE : ", metrics.mean_squared_error(y_te,y_predlr2))

MAE :  12.349278205792377
RMSE :  16.559442673854818
MSE :  274.21514166868405


In [22]:
# hyper parameter tuning for Random forest model2
from sklearn.model_selection import RandomizedSearchCV
# Number of trees in random forest
rf2 = RandomForestRegressor()
n_estimators = [int(x) for x in np.linspace(start = 200, stop = 2000, num = 10)]
# Number of features to consider at every split
max_features = ['auto', 'sqrt']
# Maximum number of levels in tree
max_depth = [int(x) for x in np.linspace(10, 110, num = 11)]
max_depth.append(None)
# Minimum number of samples required to split a node
min_samples_split = [2, 5, 10]
# Minimum number of samples required at each leaf node
min_samples_leaf = [1, 2, 4]
# Method of selecting samples for training each tree
bootstrap = [True, False]
# Create the random grid
random_grid = {'n_estimators': n_estimators,
               'max_features': max_features,
               'max_depth': max_depth,
               'min_samples_split': min_samples_split,
               'min_samples_leaf': min_samples_leaf,
               'bootstrap': bootstrap}

rf_random = RandomizedSearchCV(estimator = rf2, param_distributions = random_grid, n_iter = 100, cv = 3, verbose=2, random_state=42, n_jobs = -1)
rf_random.fit(x_te, y_te)
# to find the best params


Fitting 3 folds for each of 100 candidates, totalling 300 fits


[Parallel(n_jobs=-1)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  33 tasks      | elapsed: 19.6min
[Parallel(n_jobs=-1)]: Done 154 tasks      | elapsed: 91.4min
[Parallel(n_jobs=-1)]: Done 300 out of 300 | elapsed: 635.8min finished


RandomizedSearchCV(cv=3, error_score='raise-deprecating',
                   estimator=RandomForestRegressor(bootstrap=True,
                                                   criterion='mse',
                                                   max_depth=None,
                                                   max_features='auto',
                                                   max_leaf_nodes=None,
                                                   min_impurity_decrease=0.0,
                                                   min_impurity_split=None,
                                                   min_samples_leaf=1,
                                                   min_samples_split=2,
                                                   min_weight_fraction_leaf=0.0,
                                                   n_estimators='warn',
                                                   n_jobs=None, oob_score=False,
                                                   random_sta...


In [23]:
rf_random.best_params_

{'n_estimators': 600,
 'min_samples_split': 2,
 'min_samples_leaf': 2,
 'max_features': 'sqrt',
 'max_depth': 110,
 'bootstrap': False}

# Random forest with hyper parameter tuning

In [25]:
rfr3 = RandomForestRegressor({'n_estimators': 600,
 'min_samples_split': 2,
 'min_samples_leaf': 2,
 'max_features': 'sqrt',
 'max_depth': 110,
 'bootstrap': False})


In [28]:

rfr3 = RandomForestRegressor(n_estimators =  600,
 min_samples_split= 2,
 min_samples_leaf =2,
 max_features= 'sqrt',
 max_depth= 110,
 bootstrap = False)

In [29]:
rfrmodel3 = rfr3.fit(x_tr,y_tr)
y_pred3 = rfrmodel3.predict(x_te)
#checking the model performance by MAE,RMSE,MSE
print("MAE : ", metrics.mean_absolute_error(y_te,y_pred3))
print("RMSE : ", sqrt(metrics.mean_squared_error(y_te,y_pred3)))
print("MSE : ", metrics.mean_squared_error(y_te,y_pred3))

MAE :  4.899047390447701
RMSE :  8.794933714388995
MSE :  77.35085904049619


# Saving the model for future usage

In [36]:
import pickle 
  
# Save the trained model as a pickle string. 

filename = 'finalized_model_rf.sav'
pickle.dump(rfrmodel3, open(filename, 'wb'))
#saved_model_Rf = pickle.dumps(rfrmodel3) 
  
# Load the pickled model 
#rfr_from_pickle = pickle.loads(saved_model_rf) 
loaded_model = pickle.load(open(filename, 'rb'))
  
# Use the loaded pickled model to make predictions 
#knn_from_pickle.predict(X_test)

In [37]:
loaded_model

RandomForestRegressor(bootstrap=False, criterion='mse', max_depth=110,
                      max_features='sqrt', max_leaf_nodes=None,
                      min_impurity_decrease=0.0, min_impurity_split=None,
                      min_samples_leaf=2, min_samples_split=2,
                      min_weight_fraction_leaf=0.0, n_estimators=600,
                      n_jobs=None, oob_score=False, random_state=None,
                      verbose=0, warm_start=False)

In [ ]:
# as we saw there is an improvement in the model performance after we done hyper parameter tuning
